examples/imagenet.py

```diff
@@ -453,9 +473,14 @@ def train(train_loader, model, criterion, optimizer, epoch, args):
         top5.update(acc5[0], images.size(0))
 
         # compute gradient and do SGD step
-        optimizer.zero_grad()
         loss.backward()
-        optimizer.step()
+
+        if n_virtual_steps > 1:
+            optimizer.virtual_step()
+
+        if (i + 1) % n_virtual_steps == 0:
+            optimizer.step()
```

Comment:
are we not taking a step if len_data/batch is not dividable by effective_batch

Reply:
Here's an example of what I had in mind:
- len(train_loader) = 1250
- batch_size = 100
- effective_batch_size = 500

so n_virtual_steps = 5, and i ranges from 0 to 12

In a training epoch, we'll call .step() when (i+1)=5, and when (i+1)=10, to process batches of size 500.
Then, the last steps will process mini-batches of size 100, 100, 50, and exit the training loop.
So we're left with 3 mini-batches accumulated into param.grad_sample or param.summed_grad

Reply:
exactly, so what's happening to those in the next epoch? am I missing something?
we either have to miss them by calling zero grad or handle them at the end of the loop, correct?

Reply:
right, so in the next epoch, `i` gets reset to 0 so we'll accumulate 5 more mini-batches before we call `optimizer.step()`.
So now we have an effective batch consisting of 8 mini-batches (3 from the previous epoch, and 5 from the current epoch) and with 750 samples total. 

This is problematic for two reasons: 

1. It could cause memory issues if we're planning on holding the entire batch in memory, and we were expecting at most 500 gradients.

2. The privacy accounting for that batch is incorrect because we're assuming a sample probability of `effective_batch_size / len(train_loader) = 500/1250` whereas this batch has a sample probability of `750/1250`.

Reply:
So my thought is we should err on the conservative side (as epsilon is a an upper-bound). 
To do that I'd suggest:
either:
calling step() once at the end of the loop and checking inside step if there is un-finished business, we could check the accumulation_state or yet cleaner have a flag that tracks state. if a virtual step is called the flag is set and on virtual=false flag is cleared.
or:
do the same, but call the method finalize() [just sounds better and less confusing IMO]. For this you could simply just promote finalize_batch.
or:
handle everything internally in the engine counting number of calls to step and finalizing whenever needed. For this there would be no need for a virtual_step method at all. [I kind of like this one, but also I know this completely contradicts the talks we had offline, so may be we can do this in another diff later on [let me know what you think about this :) ]

Reply:
I guess the easiest way would be to do:
```python
if ((i + 1) % n_virtual_steps == 0) or ((i+1) == len(train_loader)):
    optimizer.step()
```

Then we'll always take a step after the last mini-batch, regardless of whether we've completed an effective batch or not.

Reply:
I like that!

Reply:
could you make this change as well please (I'd assume it would need some minimal changes in the step function as well) and we should be more than ready to land this amazing diff.

Reply:
Ok. There's still the question of when/where to delete the `param.grad_sample` fields so that they stop accumulating.
If we want to do that in `clipper.step()`, we'd have to rewrite a bunch of tests.
I think the most natural thing is still to monkeypatch `optimizer.zero_grad()` and do this there, as it aligns nicely with the semantics of `zero_grad()` for accumulation of regular gradients.

Reply:
With the latest commits, the deletions are now handled directly in `clipper.step()`, so that gradient aggregators are reset as soon as they are consumed.
And we just have to make sure to always take a step on the last mini-batch of a training epoch.